tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks/WriteSourceRepoProperties.cs

```diff
@@ -0,0 +1,260 @@
+                        // core-sdk does this - OfficialBuildId is only used for their fake package and not in anything shipped
+                        return (DateTime.Now.ToString("yyyyMMdd.1"), releaseParts[0]);
+                    }
+                    else
+                    {
+                        // NuGet does this - arbitrary build IDs
+                        return (releaseParts[1], releaseParts[0]);
+                    }
+                }
+                else if (releaseParts.Length == 3)
+                {
+                    if (int.TryParse(releaseParts[1], out int datePart) && int.TryParse(releaseParts[2], out int buildPart))
+                    {
+                        return ($"20{((datePart / 1000))}{((datePart % 1000) / 50):D2}{(datePart % 50):D2}.{buildPart}", releaseParts[0]);
+                    }
+                }
+            }
+            else
+            {
+                // finalized version number (x.y.z) - probably not our code
+                // VSTest, Application Insights, Newtonsoft.Json do this
+                return (DateTime.Now.ToString("yyyyMMdd.1"), string.Empty);
+            }
+
+            throw new FormatException($"Can't derive a build ID from version {version} (commit count {commitCount})");
+        }
+
+        private static string GetRepoNameOrDefault(Dependency dependency)
+        {
+            if (dependency.RepoName != null)
+            {
+                return dependency.RepoName;
+            }
+            var repoUrl = dependency.Uri;
+            if (repoUrl.EndsWith(".git"))
+            {
+                repoUrl = repoUrl.Substring(0, repoUrl.Length - ".git".Length);
+            }
+            return repoUrl.Substring(repoUrl.LastIndexOf("/") + 1);
+        }
+
+        private static string GetCommitCount(string gitDir, string hash)
+        {
+            return RunGitCommand($"rev-list --count {hash}", gitDir: gitDir);
+        }
+
+        private static string GetCommitDate(string gitDir, string hash)
+        {
+            return RunGitCommand($"log -1 --format=%cd --date=short", gitDir: gitDir);
+        }
+
+        private static string RunGitCommand(string command, string workTree = null, string gitDir = null)
+        {
+            // Windows Git requires these to be before the command
+            if (workTree != null)
+            {
+                command = $"--work-tree={workTree} {command}";
+            }
+            if (gitDir != null)
+            {
+                command = $"--git-dir={gitDir} {command}";
+            }
+            var psi = new ProcessStartInfo { FileName = "git", Arguments = command, RedirectStandardOutput = true, RedirectStandardError = true, UseShellExecute = false, CreateNoWindow = true };
+            var p = new Process { StartInfo = psi };
+            p.Start();
+            var output = p.StandardOutput.ReadToEnd();
+            var error = p.StandardError.ReadToEnd();
+            p.WaitForExit();
```

Comment:
Maybe consider using `Exec` task for its process capture implementation.

Reply:
Changed as suggested.